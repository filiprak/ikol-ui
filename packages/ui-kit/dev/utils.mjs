import { existsSync } from 'fs';
import { readFile, writeFile, mkdir } from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export async function useTemplateFile(filename, outputFilepath, variables) {
  const data = await readFile(path.resolve(__dirname, 'templates', filename), 'utf8');
  return useTemplate(data, outputFilepath, variables);
}

export async function useTemplate(templateStr, outputFilepath, variables) {
  const updatedTemplate = templateStr.replace(/\[\[([a-zA-Z0-9_]+)\]\]/g, (match, key) => {
    return variables[key] || match;
  });

  if (!existsSync(path.dirname(outputFilepath))) {
    await mkdir(path.dirname(outputFilepath));
  }
  await writeFile(outputFilepath, updatedTemplate);
}

export const kebabize = str => {
  return str.split('').map((letter, idx) => {
      return letter.toUpperCase() === letter
          ? `${idx !== 0 ? '-' : ''}${letter.toLowerCase()}`
          : letter;
  }).join('');
}

export const ROOT_DIR = path.resolve(__dirname, '..');
