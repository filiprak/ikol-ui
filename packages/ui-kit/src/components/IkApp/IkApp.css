body {
    position: relative;
    background-color: var(--background-color);
    color: var(--on-background-color-1);
    font-family: "Open Sans", sans-serif;
    letter-spacing: 0;
    font-weight: 600;
    font-size: 13px;
}

::selection {
    background-color: var(--primary-color) !important;
    color: var(--on-primary-color) !important;
}

::-moz-selection {
    background-color: var(--primary-color) !important;
    color: var(--on-primary-color) !important;
}

::placeholder {
    color: var(--on-background-color-4) !important;
    font-weight: 600;
}

::-moz-placeholder {
    color: var(--on-background-color-4) !important;
    font-weight: 600;
}

::-webkit-input-placeholder {
    color: var(--on-background-color-4) !important;
    font-weight: 600;
}

:-ms-input-placeholder {
    color: var(--on-background-color-4) !important;
    font-weight: 600;
}

::-ms-input-placeholder {
    color: var(--on-background-color-4) !important;
    font-weight: 600;
}

/* number input arrows */
input::-webkit-outer-spin-button,
input::-webkit-inner-spin-button {
    -webkit-appearance: none;
    margin: 0;
}

input[type=number] {
    -moz-appearance: textfield;
}

.ik-app {
    display: grid;
    align-content: start;
    position: relative;
    -webkit-backface-visibility: hidden;
    backface-visibility: hidden;
    min-height: 100%;
    max-width: 100%;
}

.ik-app--fill {
    width: 100%;
    height: 100%;
    align-content: normal;
}

.ik-app-bar {
    position: fixed;
    top: 0;
    left: 0;
    flex: 0 0 auto;
    display: flex;
    align-items: center;
    width: 100%;
    max-width: 100%;
    z-index: var(--z-index-app-bar);
    background-color: var(--background-color);
    border-bottom: 1px solid var(--surface-darken-color);
}

.ik-app-bar__back,
.ik-app-bar__burger {
    padding-left: 15px;
    font-size: 24px;
}

.ik-app-bar__title {
    max-width: 50%;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
    font-size: 18px;
    padding-left: 15px;
    min-width: 0;
}

.ik-app--mobile .ik-app-bar__title {
    max-width: 100%;
}

.ik-app-bar__content {
    flex: 1 1 auto;
    padding: 0 25px;
    min-width: 0;
}

.ik-app-content {
    display: grid;
    min-height: 0;
}

input::-webkit-contacts-auto-fill-button {
    visibility: hidden;
    display: none !important;
    pointer-events: none;
    position: absolute;
    right: 0;
}
