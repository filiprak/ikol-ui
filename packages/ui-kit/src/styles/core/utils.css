/* shadow */
.ik-shadow-0 {
    box-shadow: none;
}

.ik-shadow-1 {
    box-shadow: 0 1px 3px 0 var(--shadow-color-1);
}

.ik-shadow-2 {
    box-shadow: 0 2px 5px 0 var(--shadow-color-1);
}

.ik-shadow-3 {
    box-shadow: 0 2px 10px 0 var(--shadow-color-1);
}

.ik-shadow-4 {
    box-shadow: 0 2px 15px 0 var(--shadow-color-1);
}

.ik-shadow-5 {
    box-shadow: 0 2px 20px 0 var(--shadow-color-1);
}

.ik-shadow-6 {
    box-shadow: 0 3px 30px 0 var(--shadow-color-1);
}

/* skeleton background */
@-webkit-keyframes skeleton-loading {
    to {
        transform: translateX(100%)
    }
}

@keyframes skeleton-loading {
    to {
        transform: translateX(100%)
    }
}

.ik-bg-skeleton {
    background: var(--skeleton-color-1);
    overflow: hidden;
    position: relative;
    user-select: none;
}

.ik-bg-skeleton:after {
    background: linear-gradient(90deg, hsla(0, 0%, 100%, 0), var(--skeleton-color-2), hsla(0, 0%, 100%, 0));
    -webkit-animation: skeleton-loading 1.5s infinite;
    animation: skeleton-loading 1.5s infinite;
    content: '';
    height: 100%;
    left: 0;
    position: absolute;
    right: 0;
    top: 0;
    transform: translateX(-100%);
    z-index: 1
}

.ik-cursor--pointer {
    cursor: pointer;
}

/* design utils */
.ik-design--bg-primary {
    background-color: var(--primary-color);
    color: var(--on-primary-color);
}

.ik-design--bg-success {
    background-color: var(--success-color);
    color: var(--on-success-color);
}

.ik-design--bg-error {
    background-color: var(--error-color);
    color: var(--on-error-color);
}

.ik-design--bg-accent {
    background-color: var(--accent-color);
    color: var(--on-accent-color);
}

.ik-design--bg-secondary {
    background-color: var(--secondary-color);
    color: var(--on-secondary-color);
}

.ik-design--bg-default {
    background-color: var(--surface-darken-color);
    color: var(--on-surface-darken-color-3);
}

/* height */
.ik-height--full {
    height: 100%;
}

.ik-width--full {
    width: 100%;
}
