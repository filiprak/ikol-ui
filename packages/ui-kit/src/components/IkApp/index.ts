import IkApp from './IkApp.vue';
import IkAppContent from './IkAppContent.vue';
import IkAppBar from './IkAppBar.vue';

export interface IkAppContent extends InstanceType<typeof IkAppContent> { };
export interface IkAppBar extends InstanceType<typeof IkAppBar> { };
export interface IkApp extends InstanceType<typeof IkApp> { };

export { IkApp };
export { IkAppContent };
export { IkAppBar };
