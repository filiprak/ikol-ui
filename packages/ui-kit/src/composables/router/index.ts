import type { InjectionKey, Ref } from 'vue';
import { inject } from 'vue';
import { getCurrentInstance, logError } from '@ui/utils/core';
import { IkRouterManager } from '@ui/composables/router/IkRouterManager';
import type { IkRouteManager } from '@ui/composables/router/IkRouteManager';
import type { RouteRecordRaw } from 'vue-router';
import { createRouter as _createRouter, createWebHashHistory } from 'vue-router';

export const ROUTER_SYMBOL: InjectionKey<IkRouterManager> = Symbol.for('ik-router');

export interface IkRouterOptions {
    routes: Readonly<RouteRecordRaw[]>
}

export function createRouter(options: IkRouterOptions): IkRouterManager {
    return new IkRouterManager(
        _createRouter({
            history: createWebHashHistory(),
            routes: options.routes,
        })
    );
}

export function useRouter(): IkRouterManager {
    let router;

    if (getCurrentInstance('useRouter')) {
        router = inject(ROUTER_SYMBOL);
    }

    if (!router) {
        router = createRouter({ routes: [] });
        logError('Injection "router" not found');
    }

    return router;
}

export function useRoute(): Ref<IkRouteManager> {
    let router;

    if (getCurrentInstance('useRoute')) {
        router = inject(ROUTER_SYMBOL);
    }

    if (!router) {
        router = createRouter({ routes: [] });
        logError('Injection "router" not found');
    }

    return router.route;
}
