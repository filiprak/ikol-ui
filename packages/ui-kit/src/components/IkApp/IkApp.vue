<script lang="ts">
import '@ui/styles';
import './IkApp.css';
import { defineComponent, h, unref } from 'vue';
import { toCssUnits } from '@ui/utils/helpers';

import type { IkAppBar } from '@ui/components/IkApp';
import type { IkNavigation } from '@ui/components/IkNavigation';
import { IkModalRenderer } from '@ui/components/IkModal';
import { useDevice } from '@ui/composables/device';

export default defineComponent({
    provide() {
        return {
            app: this,
        };
    },
    props: {
        loading: {
            type: Boolean,
            default: false,
        },
        fill: {
            type: Boolean,
            default: false,
        },
    },
    data() {
        return {
            nav: null as IkNavigation | null,
            bar: null as IkAppBar | null,
        };
    },
    computed: {
        styles() {
            if (!unref(this.device.mobile)) {
                return {
                    paddingLeft: toCssUnits(this.nav_width),
                    paddingTop: toCssUnits(this.bar_height),
                };
            } else {
                return {
                    paddingTop: toCssUnits(this.bar_height),
                };
            }
        },
        nav_width() {
            if (this.nav) {
                return this.nav.mini_width;
            } else {
                return 0;
            }
        },
        bar_height() {
            if (this.bar) {
                return this.bar.height;
            } else {
                return 0;
            }
        },
    },
    methods: {
        registerBar(bar: IkAppBar) {
            this.bar = bar || null;
        },
        registerNav(nav: IkNavigation) {
            this.nav = nav || null;
        },
        unregisterBar() {
            this.bar = null;
        },
        unregisterNav() {
            this.nav = null;
        },
    },
    render() {
        let children = this.$slots.default?.() || [];

        if (this.loading) {
            children = [];
        }

        children.push(h(IkModalRenderer));

        return h('div', {
            class: {
                'ik-app': true,
                'ik-app--fill': this.fill,
                'ik-app--mobile': unref(this.device.mobile),
            },
            style: this.styles,
        }, children);
    },
    setup() { return { device: useDevice() }; },
});
</script>
