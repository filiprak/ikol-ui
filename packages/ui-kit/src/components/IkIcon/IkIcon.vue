<script setup lang="ts">
import '@ui/styles';
import './IkIcon.css';
import { type PropType, h, useAttrs } from 'vue';
import { toCssUnits, getIconClasses } from '@ui/utils/helpers';
import { useRender } from '@ui/composables/render';
import type { IkUIDesignColor } from '@ui/types/utils';

const props = defineProps({
    tag: {
        type: String,
        default: 'i',
    },
    icon: {
        type: String,
    },
    size: {
        type: [Number, String],
    },
    size_px: {
        type: [Number, String],
    },
    disabled: {
        type: Boolean,
        default: false,
    },
    circle: {
        type: Boolean,
        default: false,
    },
    design: {
        type: String as PropType<`${IkUIDesignColor}`>,
    },
    variant: {
        type: String,
    },
});
const attrs = useAttrs();

useRender(() => {
    return h(props.tag, {
        class: [
            {
                'ik-icon': true,
                ['ik-icon--' + props.design]: !!props.design,
                ['ik-icon--' + props.variant]: !!props.variant,
                'ik-icon--circle': props.circle,
                'ik-icon--clickable': !props.disabled && !!attrs.onClick,
                'ik-icon--disabled': props.disabled,
                ['fa-' + props.size + 'x']: !!props.size,
            },
            getIconClasses(props.icon || ''),
        ],
        style: {
            width: props.size_px && props.circle ? toCssUnits(props.size_px, 'px') : null,
            height: props.size_px && props.circle ? toCssUnits(props.size_px, 'px') : null,
            fontSize: props.size_px ? toCssUnits(props.circle ? (Number(props.size_px) / 2) : props.size_px, 'px') : null,
        },
    });
});
</script>
