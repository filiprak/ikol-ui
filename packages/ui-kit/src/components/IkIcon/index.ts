import IkIcon from './IkIcon.vue';
export interface IkIcon extends InstanceType<typeof IkIcon> { };
export { IkIcon };
