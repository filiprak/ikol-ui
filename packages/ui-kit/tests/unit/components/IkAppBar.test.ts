import type { RenderOptions } from '@ui/tests/tools/utils';
import { renderShallow } from '@ui/tests/tools/utils';
import { IkApp, IkAppBar } from '@ui/components/IkApp';
import { IkIcon } from '@ui/components/IkIcon';
import { getRouterMock } from '@ui/tests/mocks/router';

describe('IkAppBar', () => {
    let router = getRouterMock();

    const getInstance = (options?: RenderOptions<typeof IkAppBar>) => renderShallow(IkAppBar, {
        global: {
            provide: { app: getAppInstance() },
        },
        router,
    }, options);
    const getAppInstance = () => renderShallow(IkApp).vm;

    beforeEach(() => {
        router = getRouterMock();
    });

    it('renders with correct default state', () => {
        const wrapper = getInstance();

        expect(wrapper.classes()).toStrictEqual([
            'ik-app-bar',
        ]);

        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with correct default mobile state', () => {
        const wrapper = getInstance({
            device: { mobile: true },
        });

        expect(wrapper.classes()).toStrictEqual([
            'ik-app-bar',
        ]);

        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with custom width', () => {
        const wrapper = getInstance({
            props: { height: 77 },
        });

        expect(wrapper.attributes('style')).toContain('left: 0px; height: 77px;');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with default slot', () => {
        const wrapper = getInstance({
            slots: {
                default: 'This is app bar content...',
            },
        });

        expect(wrapper.text()).toContain('This is app bar content...');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with title slot', () => {
        const wrapper = getInstance({
            slots: {
                title: 'This is title...',
            },
        });

        expect(wrapper.text()).toContain('This is title...');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with back nav button', async () => {
        const replace = jest.spyOn(router, 'replace');

        const wrapper = getInstance({
            props: {
                back: true,
            },
            global: {
                stubs: { IkIcon },
            },
        });

        const icon = wrapper.find('.ik-app-bar__back').findComponent(IkIcon);

        await router.push('/route2');
        await icon.trigger('click');

        expect(icon.isVisible()).toBeTruthy();
        expect(replace).toHaveBeenCalledWith('/');
    });

    it('toggles nav on burger click', async () => {
        const toggleMock = jest.fn();
        const wrapper = getInstance({
            device: { mobile: true },
            global: {
                provide: { app: { nav: { toggle: toggleMock }, registerBar: jest.fn() } },
            },
        });

        await wrapper.find('.ik-app-bar__burger').findComponent(IkIcon).trigger('click');

        expect(toggleMock).toHaveBeenCalled();
    });

    it('applies title size', () => {
        const wrapper = getInstance({
            props: {
                title_size: 77,
            },
            slots: {
                title: 'This is title...',
            },
        });

        expect(wrapper.find('.ik-app-bar__title').attributes('style')).toContain('width: 77%');
    });
});
