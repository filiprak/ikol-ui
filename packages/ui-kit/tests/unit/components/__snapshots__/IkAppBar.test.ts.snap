// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IkAppBar renders with correct default mobile state 1`] = `
<header
  class="ik-app-bar"
  style="left: 0px; height: 56px;"
>
  <div
    class="ik-app-bar__burger"
  >
    <ik-icon-stub
      circle="false"
      disabled="false"
      icon="bars:light"
      tag="i"
    />
  </div>
  <div
    class="ik-app-bar__back"
    style="display: none;"
  >
    <ik-icon-stub
      circle="false"
      disabled="false"
      icon="long-arrow-left:regular"
      tag="i"
    />
  </div>
  <div
    class="ik-app-bar__title"
    style="display: none;"
  >
    <!---->
  </div>
  <div
    class="ik-app-bar__content"
  >
    <!---->
  </div>
</header>
`;

exports[`IkAppBar renders with correct default state 1`] = `
<header
  class="ik-app-bar"
  style="left: 0px; height: 56px;"
>
  <div
    class="ik-app-bar__burger"
    style="display: none;"
  >
    <ik-icon-stub
      circle="false"
      disabled="false"
      icon="bars:light"
      tag="i"
    />
  </div>
  <div
    class="ik-app-bar__back"
    style="display: none;"
  >
    <ik-icon-stub
      circle="false"
      disabled="false"
      icon="long-arrow-left:regular"
      tag="i"
    />
  </div>
  <div
    class="ik-app-bar__title"
    style="display: none;"
  >
    <!---->
  </div>
  <div
    class="ik-app-bar__content"
  >
    <!---->
  </div>
</header>
`;

exports[`IkAppBar renders with custom width 1`] = `
<header
  class="ik-app-bar"
  style="left: 0px; height: 77px;"
>
  <div
    class="ik-app-bar__burger"
    style="display: none;"
  >
    <ik-icon-stub
      circle="false"
      disabled="false"
      icon="bars:light"
      tag="i"
    />
  </div>
  <div
    class="ik-app-bar__back"
    style="display: none;"
  >
    <ik-icon-stub
      circle="false"
      disabled="false"
      icon="long-arrow-left:regular"
      tag="i"
    />
  </div>
  <div
    class="ik-app-bar__title"
    style="display: none;"
  >
    <!---->
  </div>
  <div
    class="ik-app-bar__content"
  >
    <!---->
  </div>
</header>
`;

exports[`IkAppBar renders with default slot 1`] = `
<header
  class="ik-app-bar"
  style="left: 0px; height: 56px;"
>
  <div
    class="ik-app-bar__burger"
    style="display: none;"
  >
    <ik-icon-stub
      circle="false"
      disabled="false"
      icon="bars:light"
      tag="i"
    />
  </div>
  <div
    class="ik-app-bar__back"
    style="display: none;"
  >
    <ik-icon-stub
      circle="false"
      disabled="false"
      icon="long-arrow-left:regular"
      tag="i"
    />
  </div>
  <div
    class="ik-app-bar__title"
    style="display: none;"
  >
    <!---->
  </div>
  <div
    class="ik-app-bar__content"
  >
    
    This is app bar content...
    
  </div>
</header>
`;

exports[`IkAppBar renders with title slot 1`] = `
<header
  class="ik-app-bar"
  style="left: 0px; height: 56px;"
>
  <div
    class="ik-app-bar__burger"
    style="display: none;"
  >
    <ik-icon-stub
      circle="false"
      disabled="false"
      icon="bars:light"
      tag="i"
    />
  </div>
  <div
    class="ik-app-bar__back"
    style="display: none;"
  >
    <ik-icon-stub
      circle="false"
      disabled="false"
      icon="long-arrow-left:regular"
      tag="i"
    />
  </div>
  <div
    class="ik-app-bar__title"
  >
    
    This is title...
    
  </div>
  <div
    class="ik-app-bar__content"
  >
    <!---->
  </div>
</header>
`;
